### ISSUE TITLE
"Error reading BEARER_TOKEN_FILE: permission denied" when deploying docker-registry

### ISSUE THREAD

**Issue by sjenning** [openshift/origin#6981]
I'm running v1.1.1.1 on Rawhide with Docker 1.10.0-rc2.  I am running into this when I try to deploy the docker-registry:

```
E0202 21:25:01.768612       1 clientcmd.go:176] Error reading BEARER_TOKEN_FILE "/var/run/secrets/kubernetes.io/serviceaccount/token": open /var/run/secrets/kubernetes.io/serviceaccount/token: permission denied
E0202 21:25:01.776267       1 clientcmd.go:194] Error reading BEARER_TOKEN_FILE "/var/run/secrets/kubernetes.io/serviceaccount/token": open /var/run/secrets/kubernetes.io/serviceaccount/token: permission denied
F0202 21:25:01.840429       1 deployer.go:65] couldn't get deployment default/docker-registry-1: User "system:anonymous" cannot get replicationcontrollers in project "default"
```

docker inspect on the origin-deployer container:

```
        "Mounts": [
            {
                "Source": "/var/lib/origin/openshift.local.volumes/pods/80c93db2-c9f5-11e5-9c3f-fa163e61036e/volumes/kubernetes.io~secret/deployer-token-qj1bv",
                "Destination": "/var/run/secrets/kubernetes.io/serviceaccount",
                "Mode": "ro,Z",
                "RW": false,
                "Propagation": "rprivate"
            },
...
        ],
...
                "BEARER_TOKEN_FILE=/var/run/secrets/kubernetes.io/serviceaccount/token",
```

The token file on the host exists and is world readable.

```
# ls -l /var/lib/origin/openshift.local.volumes/pods/80c93db2-c9f5-11e5-9c3f-fa163e61036e/volumes/kubernetes.io~secret/deployer-token-qj1bv/token
-r--r--r--. 1 root root 850 Feb  2 21:39 /var/lib/origin/openshift.local.volumes/pods/80c93db2-c9f5-11e5-9c3f-fa163e61036e/volumes/kubernetes.io~secret/deployer-token-qj1bv/token
```

v1.1.1.1 worked fine on F23 with Docker 1.9.1.  Possibly an issue introduced with root mount propagation in Docker 1.10?

I disabled selinux, but the problem persists.

**Comment by sjenning** [comment]
Further investigation shows that in docker 1.9, /run/secrets (/var/run is a symlink to /run) is a rw tmpfs, but in docker 1.10 it is a ro bind mount with mode 700, hence my permission denied error.

Volume mounts inside /run/secrets are not being honored.

It seems that docker has created there own mechanism for putting secrets into the container; however, I can find zero documentation about it.

**Comment by sjenning** [comment]
Clarification, docker is mounting _over_ /var/secrets:

```
# findmnt
TARGET                                SOURCE                                                                           FSTYPE OPTIONS
...
|-/run/secrets/kubernetes.io/serviceaccount
|                                     tmpfs                                                                            tmpfs  ro,relatime,rootcontext="system_u:object_r:svirt_sandbox_file_t:s0:c113,c301",seclabel
...
`-/run/secrets                        /dev/vda1[/var/lib/docker/containers/ecbff5bb7723291f6ff8b1b0e43d460ab6c4368b6fd0eb0579acb30c89351220/secrets]
                                                                                                                       ext4   ro,relatime,seclabel,data=ordered
```

**Comment by sjenning** [comment]
The reason I could find documentation on this is because the Fedora docker packages are not pulled from upstream docker; they are pulled from a forked repo here: 
https://github.com/projectatomic/docker

This patch is the issue I think:
https://github.com/projectatomic/docker/commit/202463d634484b52246995628eca6fd734fd65fb

**Comment by smarterclayton** [comment]
@mrunalp @rhatdan docker 1.10 fedora is mounting over the kubernetes secret path, which breaks kubernetes and Openshift.  I assume this is the docker secrets patch?

**Comment by mrunalp** [comment]
Yes, we could detect that it is already mounted and skip it to prevent this issue.

**Comment by rhatdan** [comment]
/var/run/secrets should be mounted before any user volumes are mounted.

**Comment by smarterclayton** [comment]
Thanks for the lightning fast fix Dan

**Comment by sjenning** [comment]
I can verify that this fix helps

```
# findmnt
...
|-/run/secrets                               /dev/vda1[/var/lib/docker/containers/3839e8a09faef90c7d6ef8be15e01413352299a46df840094141af2d079c1b4f/secrets]
|                                                                                                                             ext4   rw,relatime,seclabel,data=ordered
| `-/run/secrets/kubernetes.io/serviceaccount
|                                            tmpfs                                                                            tmpfs  ro,relatime,rootcontext="system_u:object_r:svirt_sandbox_file_t:s0:c349,c948",seclabel

# ls /run/secrets/kubernetes.io/serviceaccount/token 
/run/secrets/kubernetes.io/serviceaccount/token
```

But there is still an issue: /run/secrets is mode 700.  The deployer uses a not root user to do the work so I'm still getting permission denied.

With docker 1.9 /run/secrets was tmpfs mounted with mode 777.

**Comment by sjenning** [comment]
Clarification, the secret directory in the host is 700 so it is the same in the container.

This line in the patch:
```
if err := os.MkdirAll(secretsPath, 0700); err != nil {
```

is setting the mode.

For the purposes RHEL uses this for, it makes sense.  But not for kube.

**Comment by sjenning** [comment]
@rhatdan I can confirm that this modification fixes this issue completely for me.  I'm not sure if it is the correct way, but it is _a_ way.

https://github.com/sjenning/docker/commit/27f709fd6c383f7f5a7804188ad1aa5ffe888d5e

**Comment by rhatdan** [comment]
It should never have been 777, this would be a security issue.  If necessary we could make it 755, but I don't think you actually need it to be accessible by non root.

**Comment by rhatdan** [comment]
SHould be fixed in rc4.

**Comment by smarterclayton** [comment]
As long as /run/secrets/kubernetes.io doesn't get made to be 700 we should
be ok.

**Comment by rhatdan** [comment]
I am not sure about that, if a non priv process needs to get to /run/secretes/kubernetes.io, they won't be able to since they will not be able to search through /run/secrets.
```
# docker run -ti --rm fedora sh
useradd dwalsh
sh-4.3# su dwalsh
[dwalsh@a66472a69562 run]$ cd /run/secrets/
bash: cd: /run/secrets/: Permission denied
```

**Comment by rhatdan** [comment]
I can change the permissions to 755 to fix this.

**Comment by smarterclayton** [comment]
Hrm - yeah, probably would be better for users to be able to see
/run/secrets.

Thanks

**Comment by sjenning** [comment]
Not sure which version the fix went in but I can confirm that it is fixed as of docker-1.10.3-24.git29066b4
